### ISSUE TITLE
KeePassXC-Browser and locked databases

### ISSUE THREAD

**Issue by PhrozenByte** [keepassxreboot/keepassxc-browser#397]
## Expected Behavior
1. Open a website with a login form and press the addon's button. The modal shows `Timeout or not connected to KeePassXC` (the database is locked).

2. Switch to KeePassXC and unlock the database just regularly.

3. Switch back to Firefox/Chrome/…, click the login form and all credentials are shown in the drop-down menu.

## Current Behavior
1. Open a website with a login form and press the addon's button. The modal shows `Timeout or not connected to KeePassXC` (the database is locked).

2. Switch to KeePassXC and unlock the database just regularly.

3. Switch back to Firefox/Chrome/…, click the login form. *Nothing happens.* Instead of the addon's drop-down menu, Firefox's/Chrome's/… default auto-completion drop-down is shown.

4. Click the addon's button, the modal still shows `Timeout or not connected to KeePassXC`. Weirdly enough the addon's console (see below) shows `Error 5: Timeout or not connected to KeePassXC`, even though the addon is obviously not really trying to connect to the database?

5. Click the `Reload` button in the addon's modal, the modal now shows `KeePassXC-Browser is connected to KeePassXC`. The addon's console (see below) yields a successful connection.

6. a) Click the login form and all credentials are shown in the drop-down menu.

   b) Click the login form. *Nothing happens.* Instead of the addon's drop-down menu, Firefox's/Chrome's/… default auto-completion drop-down is shown. Nothing is shown in the addon's console (see below).

      1. Click the addon's button, open `Settings` page, close `Settings` page.
      2. Click the login form and all credentials are shown in the drop-down menu.

## Steps to Reproduce (for bugs)
Case 6. a) can be easily reproduced by replicating the current behavior as shown above.

Unfortunately I can't reliably reproduce case 6. b). It happens pretty often, but I couldn't find the exact circumstances that differentiate case 6. a) and b).

I also checked the addon's console (Developer Tools), nothing special in there (12:25:39 is step 1, 12:26:00 is step 4, 12:26:01 and 12:26:02 are both step 5):

```
12:25:39.413 Error 5: Timeout or not connected to KeePassXC. keepass.js:966:5
12:25:39.991 Error: Could not establish connection. Receiving end does not exist. background.js:1:16542 
12:26:00.718 Error 5: Timeout or not connected to KeePassXC. keepass.js:966:5
12:26:01.733 Connecting to native messaging host org.keepassxc.keepassxc_browser keepass.js:882:5
12:26:02.251 Server public key: ***
```

## Possible Solution
As soon as a login form is clicked, KeePassXC-Browser should try to reload the database automatically. It might be reasonable to limit the number of auto-reloads to e.g. once per second.

## Debug info
KeePassXC - 2.3.1
KeePassXC-Browser - 1.3.2
Operating system: Ubuntu 18.04 (*bionic*)
Browser: Firefox 64.0

**Comment by droidmonkey** [comment]
This is because it is a known bug in version 2.3.1. This has been fixed and is available in 2.3.4.

**Comment by varjolintu** [comment]
The current version of the extension doesn't support auto-connecting (yet) but there's already a pull request available. The minimum version of KeePassXC for this must be at least 2.3.4 just like said above.

**Comment by PhrozenByte** [comment]
Sorry for the late response @droidmonkey @varjolintu, unfortunately I still experience this issue after upgrading to KeePassXC 2.3.4 (using the PPA).

**Comment by PhrozenByte** [comment]
@droidmonkey @varjolintu Shall I open a new Issue or do you want to re-open this one? I can't re-open it myself :wink:

**Comment by droidmonkey** [comment]
So you are saying the EXACT same behavior occurs in 2.3.4? There are times when the extension needs to be refreshed, but certainly not every time you use it.

**Comment by PhrozenByte** [comment]
Yes, the exact same behaviour occurs with KeePassXC 2.3.4, too. As described above, it doesn't happen all the time, rather randomly; I still couldn't find the exact circumstances to safely reproduce it. It happens both if KeePassXC isn't running at all, and if KeePassXC is running, but the database is locked. However, it "feels" like it happens more often if KeePassXC isn't running at all.

Here's the log from `about:debugging` when KeePassXC wasn't running:

```
13:33:13.711 Error: Could not establish connection. Receiving end does not exist.
background.js:1:16677
13:33:18.504 Error 5: Timeout or not connected to KeePassXC. keepass.js:966:5
13:33:18.558 Error: Could not establish connection. Receiving end does not exist.
background.js:1:16677
13:33:31.401 Error 5: Timeout or not connected to KeePassXC.
keepass.js:966:5
13:34:13.970 Connecting to native messaging host org.keepassxc.keepassxc_browser keepass.js:882:5
13:34:14.515 Server public key: *** keepass.js:639:17
13:34:28.465 Error: Could not establish connection. Receiving end does not exist.
```

And this is what I'm doing:

1. `13:33:13.711`: Open a empty tab
2. `13:33:18.504` and `13:33:18.558`: Browse to facebook.com
3. (not in the logs): I click the username field, nothing happens (as expected).
4. `13:33:31.401`: I open the addon's modal, it tells me "Timeout or not connected to KeePassXC" (as expected).
5. (not in the logs): I start KeePassXC and unlock my database.
6. (not in the logs): I click the username field, nothing happens (the currently expected behaviour according to @varjolintu).
7. (not in the logs): I open the addon's modal, it still tells me "Timeout or not connected to KeePassXC".
8. `13:34:13.970` and `13:34:14.515`: I click the "Reload" button in the addon's modal.
9. (not in the logs): I click the username field, nothing happens (now it's definitly a bug).
10. `13:34:28.465`: I switch to a different browser tab. Note that KeePassXC is still running and the database is still unlocked.

With KeePassXC 2.3.1 it was possible to open the addon's Settings page via the addon's modal after step 9 (nothing is shown in the logs). After doing so, it worked with KeePassXC 2.3.1. However, since I upgraded to KeePassXC 2.3.4, even this workaround doesn't work anymore. I'll have to lock the database and reopen it again, it usually (but not always) works then.

![bildschirmfoto von 2019-01-31 13-48-02](https://user-images.githubusercontent.com/920356/52055144-dad42300-255e-11e9-8f5a-aef805711a96.png)

**Comment by droidmonkey** [comment]
The only bug in that sequence is at step 9, and I agree that once a connection is made there should be a "refresh" of the credentials search so that the drop-down appears. Also, make sure you have your URL's set correctly for your entries. You MUST include "https://", using "http://" will not work for SSL websites.

**Comment by PhrozenByte** [comment]
> The only bug in that sequence is at step 9, and I agree that once a connection is made there should be a "refresh" of the credentials search so that the drop-down appears.

I'm confused tbh :confused: So, you can reproduce it? Or is it even expected behaviour? Because at the moment I don't even know how to make it work... I'll have to lock and unlock the database again?

> Also, make sure you have your URL's set correctly for your entries. You MUST include "https://", using "http://" will not work for SSL websites.

That's the case. As I said before, it is failing rather randomly, i.e. sometimes a website that worked before stops working and vice-versa.

**Comment by droidmonkey** [comment]
I am just generally confused by what you think the issue is. I see a bunch of things that could be perceived as an issue. Please clarify.

**Comment by PhrozenByte** [comment]
Basically that KeePassXC-Browser often, *not always*, doesn't work at all :laughing:

If KeePassXC isn't running, I expect KeePassXC-Browser to tell me that KeePassXC isn't running. It does, so no issue here.

After starting KeePassXC and unlocking my database, I expect KeePassXC-Browser to do what it's supposed to do: After clicking in the username field I expect KeePassXC-Browser to list all known logins for this website. I understand that this doesn't work automatically right now due to https://github.com/keepassxreboot/keepassxc-browser/issues/397#issuecomment-456793502, so no issue here.

Thus I click the "Reload" button manually. However, KeePassXC-Browser still doesn't work, even though the addon's modal tells me that it has successfully connected to KeePassXC - *and this is the issue*! No logins are shown. Nothing. There's no workaround to make KeePassXC show the logins. I have to manually lock and unlock my database. After doing so, all known logins for this website are shown correctly (most of the time at least, not always). So, definitely nothing wrong with "https://" or the stored URLs in general, otherwise it would fail all the time, not just randomly.

KeePassXC-Browser just randomly refuses to show any logins, even though there are matching logins in the database, the database is unlocked and KeePassXC is running. It should work. But doesn't.

I've no idea how to debug this, the only information I can provide is what KeePassXC-Browser prints in `about:debugging` and that's virtually nothing. Is there some sort of debug mode?

**Comment by droidmonkey** [comment]
You can simply refresh the page to get the credentials to be retrieved again. I opened a new issue based on your last message.

**Comment by PhrozenByte** [comment]
The problem is that this unfortunately doesn't work and the credentials aren't shown after refreshing the page...

**Comment by ams001** [comment]
I am still seeing this issue with keepassxc 2.5.3 and KeePassXC-Browser 1.5.4 on Chrome 79.0.3945.130 (Linux)  I can resolve the issue by simply disabling the KeePassXC-Browser plugin and re-enabling it.

**Comment by varjolintu** [comment]
@ams001 Using the Reload-button from the popup doesn't help? Could you write down the exact situation when this happens?

**Comment by ams001** [comment]
> @ams001 Using the Reload-button from the popup doesn't help? 

No.  I also tried restarting keepassxc.

> Could you write down the exact situation when this happens?
I wish I could - it appears to occur inconsistently.  Most often when I start my days work and need to re-login (unlock terminal, re-login to sites etc), but that could just be when its used most.  Restarting chrome obviously also works, just this morning I had too much state to go that route and found this report, tried the reload etc (none of which worked) and did the disable/re-enable when noting the version of KeePassXC-Browser.

**Comment by varjolintu** [comment]
This would definitely need some debugging. Is your computer logged off or sleeping before this happens?

**Comment by ams001** [comment]
The screen is locked - that's all.  keepassxc is running, database IS initially locked (due to screen lock), so the first time the message is correct, but once unlocked KeePassXC-Browser still reports database as locked despite reload/refresh etc attempts. 

I have not been able to reproduce the issue on demand (locking screen, unlocking, ...) so the issue _may_ be linked to the manner or duration of the screen or database lock, but as mentioned it may just appear this way because this is when I encounter the issue most.
